### PR TITLE
Refactor taggings

### PR DESCRIPTION
**Pull request by northeastprince**

### CODE REVIEW

File: app/models/concerns/taggable.rb [hunk 1]
```diff
@@ -6,32 +6,33 @@ module Taggable
     has_many :tags, through: :taggings
   end
 
-  def tag_as(name)
+  def tag_with(tags = nil, name: nil)
```
Comment (garyhtou) [inline app/models/concerns/taggable.rb:9]: i think this might be a little cleaner (it's untested tho)
```ruby
def tag_with(tags = [], name: nil)
  tags = tags.to_a
  raise ArgumentError unless tags.any? || name

  tags << Tag.find_by(name:) if name
  tags.each do |tag|
    taggings.create_or_find_by! tag: tag
  end
end

def tag_with!(tags = [], name: nil)
  tags = tags.to_a
  tags << Tag.find_or_create_by! name: if name

  tag_with tags
end
```

Comment (northeastprince) [reply app/models/concerns/taggable.rb:9]: This allows for use like `#tag_with Tag.first, name: "(another tag)"` which would lead to more confusion. Maybe we just allow for a `tags_or_name`.

File: app/models/concerns/taggable.rb [hunk 2]
```diff
@@ -6,32 +6,35 @@ module Taggable
...
-    if (tag = Tag.find_by(name: name))
-      tag_with tag
+  def tag_with(tags_or_names)
+    Array(tags_or_names).each do |tag_or_name|
+      if tag_or_name.is_a? Tag
+        taggings.create_or_find_by! tag: tag_or_name
+      else
+        tag = Tag.find_by name: tag_or_name
+        taggings.create_or_find_by!(tag:) if tag
+      end
```
Comment (garyhtou) [inline app/models/concerns/taggable.rb:16]:
```suggestion
    Array(tags_or_names).each do |t|
      t = Tag.find_by name: t unless t.is_a? Tag
      taggings.create_or_find_by!(tag: t) if t
```

File: app/models/concerns/taggable.rb [hunk 3]
```diff
@@ -6,32 +6,35 @@ module Taggable
...
-  def untag_as(name)
-    if (tag = Tag.find_by(name:))
-      untag tag
+  def tag_with!(tags_or_names)
+    Array(tags_or_names).each do |tag_or_name|
+      if tag_or_name.is_a? Tag
+        tag_with tag_or_name
+      else
+        tag_with Tag.find_or_create_by!(name: tag_or_name)
+      end
```
Comment (garyhtou) [inline app/models/concerns/taggable.rb:26]:
```suggestion
    Array(tags_or_names).each do |t|
      t = Tag.find_or_create_by!(name: t) unless t.is_a? Tag
      tag_with t
```

File: app/models/concerns/taggable.rb [hunk 4]
```diff
@@ -6,32 +6,35 @@ module Taggable
...
-    Array(tags).each do |tag|
-      taggings.destroy_by tag:
+  def untag(tags_or_names)
+    Array(tags_or_names).each do |tag_or_name|
+      if tag_or_name.is_a? Tag
+        taggings.destroy_by tag: tag_or_name
+      else
+        tag = Tag.find_by name: tag_or_name
+        taggings.destroy_by tag: tag
+      end
```
Comment (garyhtou) [inline app/models/concerns/taggable.rb:37]:
```suggestion
    Array(tags_or_names).each do |t|
      t = Tag.find_by name: t unless t.is_a? Tag
      taggings.destroy t
```

File: test/models/tagging_test.rb [hunk 5]
```diff
@@ -1,73 +1,78 @@
...
+
+  test "checking for tags" do
+    assert @hackathon.tags.none?
+
+    @hackathon.taggings.create! tag: Tag.new(name: "Tag")
+
+    assert @hackathon.tagged_with? "Tag"
+    assert @hackathon.tagged_with? Tag.last
+
+    assert_not @hackathon.tagged_with? "Nonexistent"
```
Comment (garyhtou) [inline test/models/tagging_test.rb:16]:
```suggestion
    assert_not @hackathon.tagged_with? "Nonexistent"
    assert_not @hackathon.tagged_with? nil
```

**Review by garyhtou — APPROVED** [review]
amazing! great idea for adding `tagged_with?`

File: test/models/tagging_test.rb (hunk 5 above)
Comment (northeastprince) [reply test/models/tagging_test.rb:16]: What would this do?

**Comment by northeastprince** [comment]
Sorry, last addition 😆

File: test/models/tagging_test.rb (hunk 5 above)
Comment (garyhtou) [reply test/models/tagging_test.rb:16]: We should test of edge cases, such as passing `nil` to `tagged_with?` should returns false

**Review by garyhtou — APPROVED** [review]
let's merge it!

File: app/models/concerns/taggable.rb [hunk 6]
```diff
@@ -4,34 +4,41 @@ module Taggable
   included do
     has_many :taggings, as: :taggable, dependent: :destroy
     has_many :tags, through: :taggings
-  end
 
-  def tag_as(name)
-    if (tag = Tag.find_by(name: name))
-      tag_with tag
-    end
+    # scope that handles tags or names of tags passed in
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:8]:
```suggestion
```

File: app/models/concerns/taggable.rb [hunk 7]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless tag.is_a? Tag
+      tags.include? tag
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:19]:
```suggestion
      tags.include? tag if tag
```

File: app/models/concerns/taggable.rb [hunk 8]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless tag.is_a? Tag
+      tags.include? tag if tag
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:19]:
```suggestion
      tags.include? tag
```

File: app/models/concerns/taggable.rb [hunk 9]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+
+      joins(:taggings).where(taggings: {tag: tags})
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:17]:
```suggestion
    Array(tags_or_names).compact.all? do |tag|
```

File: app/models/concerns/taggable.rb [hunk 10]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+
+      joins(:taggings).where(taggings: {tag: tags})
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).compact.all? do |tag|
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:17]:
```suggestion
    Array(tags_or_names).all? do |tag|
```

File: app/models/concerns/taggable.rb [hunk 11]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+      tag = Tag.find_by(name: tag) unless tag.is_a? Tag
+      tags.include? tag
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:19]:
```suggestion
      tag ? tags.include?(tag) : false
```

File: app/models/concerns/taggable.rb [hunk 12]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless tag.is_a? Tag
+      tag ? tags.include?(tag) : false
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:19]:
```suggestion
      tags.include?(tag)
```

File: app/models/concerns/taggable.rb [hunk 13]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+      joins(:taggings).where(taggings: {tag: tags})
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless tag.is_a? Tag
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:18]:
```suggestion
      tag = Tag.find_by(name: tag) unless !tag || tag.is_a? Tag
```

File: app/models/concerns/taggable.rb [hunk 14]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+      joins(:taggings).where(taggings: {tag: tags})
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless !tag || tag.is_a? Tag
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:18]:
```suggestion
      tag = Tag.find_by(name: tag) unless !tag || tag.is_a?(Tag)
```

File: app/models/concerns/taggable.rb [hunk 15]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless !tag || tag.is_a?(Tag)
+      tags.include?(tag)
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:19]:
```suggestion
      tag ? tags.include?(tag) : false
```

File: app/models/concerns/taggable.rb [hunk 16]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+      joins(:taggings).where(taggings: {tag: tags})
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless !tag || tag.is_a?(Tag)
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:18]:
```suggestion
      tag = Tag.find_by(name: tag) unless tag || tag.is_a?(Tag)
```

File: app/models/concerns/taggable.rb [hunk 17]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless tag || tag.is_a?(Tag)
+      tag ? tags.include?(tag) : false
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:19]:
```suggestion
      tags.include?(tag)
```

File: app/models/concerns/taggable.rb [hunk 18]
```diff
@@ -4,34 +4,40 @@ module Taggable
...
+      joins(:taggings).where(taggings: {tag: tags})
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      tag = Tag.find_by(name: tag) unless tag || tag.is_a?(Tag)
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:18]:
```suggestion
      tag = Tag.find_by(name: tag) unless tag.is_a? Tag
```

File: app/models/concerns/taggable.rb (hunk 9 above)
Comment (northeastprince) [inline app/models/concerns/taggable.rb:17]:
```suggestion
    Array(tags_or_names).all? do |tag|
      return false unless tag
```

File: app/models/concerns/taggable.rb [hunk 19]
```diff
@@ -4,34 +4,41 @@ module Taggable
...
+      joins(:taggings).where(taggings: {tag: tags})
+    }
   end
 
-  def tag_as!(name)
-    tag = Tag.find_or_create_by! name: name
-    tag_with tag
+  def tagged_with?(tags_or_names)
+    Array(tags_or_names).all? do |tag|
+      return false unless tag
```
Comment (northeastprince) [inline app/models/concerns/taggable.rb:18]:
```suggestion
      return false unless tag.present?
```
